### PR TITLE
wallet2: fix --generate-from-json in RPC mode

### PR DESCRIPTION
**Pull request by moneromooo-monero**
The daemon address was initialized too late

### CODE REVIEW

**Comment by valuead** [comment]
thanks guys, glad we catch a bug... 
Please advise, I see some fixing was made.  Question: How to get a new version or I need to wait for a new release? 
 when it that? need to move forward with my app

**Comment by moneromooo-monero** [comment]
You can either apply the patch manually (since it's very small), or with the patch tool. You can also pull the PR from github with git (instructions in the github help FAQ, I don't recall the exact command). This will be in next release, I don't know when though.

**Comment by NanoAkron** [comment]
To merge a single PR:

First create a new branch with `git checkout -b <new branch name>`

Then `git pull https://github.com/monero-project/monero/ +refs/pull/1976/head`

**Comment by valuead** [comment]
trying to start wallet from JAVA and fails with error 1 
We downloaded all the changes and recompile. now we can work with JSON 
Another problem we can't execute code:
ProcessBuilder builder = new ProcessBuilder("C:\\dev\\monero-wallet-rpc.exe", "--rpc-bind-ip 0.0.0.0", "--rpc-bind-port 18082", "--confirm-external-bind",
       "--disable-rpc-login", "-generate-from-json \"C:\\Xmr\\1.txt\"");

final Process process = builder.start();
int errorCode = process.waitFor(); //Return 1 here 
if (errorCode!=0) {
InputStream is = process.getInputStream();
 InputStreamReader isr = new InputStreamReader(is);
 BufferedReader br = new BufferedReader(isr);
 String line;
 while ((line = br.readLine()) != null) {
   System.out.println(line);
 }
}
System.out.println("Program terminated!");

**Comment by moneromooo-monero** [comment]
--generate-from-json (missing -)

**Comment by DmRomantsov** [comment]
@moneromooo-monero Thank you for answer. 
Added - in the parameter and all characters are escaped. The same result. It start "ping", "curl.exe", "monero-wallet-cli.exe" commands successfully with the same code. But Rpc does not open and crash with code 1. The Windows documentation says that 1 is an ERROR_INVALID_FUNCTION Incorrect function.

File: src/wallet/wallet2.cpp [hunk 1]
```diff
@@ -355,7 +357,7 @@ std::unique_ptr<tools::wallet2> generate_from_json(const std::string& json_file,
       return false;
     }
 
-    wallet.reset(new tools::wallet2(testnet, restricted));
+    wallet.reset(make_basic(vm, opts).release());
```
Comment (vtnerd) [inline src/wallet/wallet2.cpp]: Assignment actually works in this situation; `wallet = make_basic(vm_opts);`. The current version should have identical behavior FWIW.

File: src/wallet/wallet2.cpp [hunk 2]
```diff
@@ -467,7 +469,7 @@ boost::optional<password_container> wallet2::password_prompt(const bool new_pass
 std::unique_ptr<wallet2> wallet2::make_from_json(const boost::program_options::variables_map& vm, const std::string& json_file)
 {
   const options opts{};
```
Comment (vtnerd) [inline src/wallet/wallet2.cpp]: This object could be created in `generate_from_json` instead to save a parameter argument.

File: src/wallet/wallet2.cpp (hunk 1 above)
Comment (moneromooo-monero) [inline src/wallet/wallet2.cpp]: That's a good question. I think it's the JSON data that takes precedence, but not 100% sure. Not something the user ought to do with conflicting parameters anyway.

**Pull request merged**